include/lbann/execution_contexts/execution_context.hpp

```diff
@@ -0,0 +1,151 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2016, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef LBANN_EXECUTION_CONTEXT_HPP
+#define LBANN_EXECUTION_CONTEXT_HPP
+
+#include "lbann/base.hpp"
+#include "lbann/comm.hpp"
+#include "lbann/io/persist.hpp"
+#include "lbann/utils/threads/thread_pool.hpp"
+#include <cereal/types/utility.hpp>
+
+namespace lbann {
+
+// Forward-declare this.
+class trainer;
+
+struct termination_criteria {
```

Comment:
I think `criteria` is better than `criterion` because there may be multiple factors (e.g., stop when some tolerance on accuracy is met *or* a hard epoch limit is hit). When the quantity is unknown, the plural is generally used.

Reply:
How do you distinguish between one `criteria` object and multiple? Criterias? And logically, I'd say that a complicated criterion that uses multiple factors is still a singular object, in the same way that a complicated logical statement with a lot of if/and statements still evaluates to a single boolean in the end.

Reply:
"There are multiple `criteria` objects." (much like "there are several `weights` objects in a `model` object" -- the choice to attempt to pluralize a type is colloquial convenience and neither syntactic nor semantic necessity)? That said, I don't care that much as long as its consistent.

I think more important than the name is making this more OO-friendly. In particular, it should have a `virtual bool check(/* something */) const = 0` member function so users don't need to know or care about the particulars of an abstract `termination_criteri*` object to know if they should be stopping. My guess is that this should be something like:
```
virtual bool check(execution_context const&) const = 0;
```